src/dhtproto/client/request/internal/GetHashRange.d

```diff
@@ -208,229 +193,195 @@ public struct GetHashRange
-            }
-        );
-        auto shared_resources = SharedResources.fromObject(
-            this.context.shared_resources);
-        shared_resources.node_hash_ranges.updateNodeHashRange(
-            this.conn.remote_address, min, max);
-
-        return State.Receiving;
+        auto supported = cast(SupportedStatus)code;
+        return GetHashRange.handleSupportedCodes(supported,
+            this.context, this.conn.remote_address);
     }
 
     /***************************************************************************
 
-        Default running state. Receives and handles hash-range update messages
-        from the node.
-
-        Next state:
-            - Receiving again
-
-        Returns:
-            next state
+        Handler policy, called from AllNodesRequest template to run the
+        request's main handling logic.
 
     ***************************************************************************/
 
-    private State stateReceiving ( )
+    private void handle ( )
     {
-        bool msg_error;
+        // Receive and store node's current hash range
+        hash_t min, max;
         this.conn.receive(
-            ( in void[] const_payload )
+            ( in void[] payload )
             {
-                Const!(void)[] payload = const_payload;
-
-                auto msg_type =
-                    *this.conn.message_parser.getValue!(MessageType)(payload);
-
-                with ( MessageType ) switch ( msg_type )
-                {
-                    case NewNode:
-                        typeof(AddrPort.naddress) addr;
-                        typeof(AddrPort.nport) port;
-                        hash_t min, max;
-                        this.conn.message_parser.parseBody(payload,
-                            addr, port, min, max);
-
-                        auto shared_resources = SharedResources.fromObject(
-                            this.context.shared_resources);
-                        shared_resources.node_hash_ranges.updateNodeHashRange(
-                            AddrPort(addr, port), min, max);
-                        break;
-
-                    case ChangeHashRange:
-                        hash_t new_min, new_max;
-                        this.conn.message_parser.parseBody(payload,
-                            new_min, new_max);
-
-                        auto shared_resources = SharedResources.fromObject(
-                            this.context.shared_resources);
-                        shared_resources.node_hash_ranges.updateNodeHashRange(
-                            this.conn.remote_address, new_min, new_max);
-                        break;
-
-                    default:
-                        log.error("Unknown message code {} received", msg_type);
-                        this.conn.shutdownWithProtocolError("Message parsing error");
-                }
+                Const!(void)[] payload_slice = payload;
```

Comment:
Out of curiosity, why is this needed? Is the `payload` not a lvalue (since `getValue` will modify it?). Would it work if you made `(in Const!(void)[] payload)`?

Reply:
It's because `MessageParser.getValue` modifies the slice that is passed to it. It's not possible to mutate the `in void[] payload`.

Reply:
That's right, but what about if the delegate would accept `Const!(void)[] payload`? I see no reason to accept it as `in`.

Reply:
Oh I see what you mean. Yeah, modifying the content of the payload should be disallowed, but modifying the slice to the payload buffer is fine. This would be a breaking change in swarm, though :(

Reply:
I love breaking changes ;-). All right, I'll create issue to target v6.0.0 in swarm.